contracts/MultiPartyEscrow.sol

```diff
@@ -0,0 +1,270 @@
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+contract MultiPartyEscrow {
+    
+    //it seems we don't need SafeMath
+   //using SafeMath for uint256;
+    
+
+    //TODO: we could use uint64 for replica_id and nonce (it could be cheaper to store but more expensive to operate with)
+
+    //the full ID of "atomic" payment channel = "[this, channel_id, nonce]"
+    struct PaymentChannel {
+        address sender;      // The account sending payments.
+        address recipient;    // The account receiving the payments.
+        uint256 replica_id;  // id of particular service replica
+        uint256 value;       // Total amount of tokens deposited to the channel. 
+        uint256 nonce;       // "nonce" of the channel (by changing nonce we effectivly close the old channel ([this, channel_id, old_nonce])
+                             //  and open the new channel [this, channel_id, new_nonce])
+                             //!!! nonce also prevents race conditon between channel_claim and channel_extend_and_add_funds 
+        uint256 expiration;  // Timeout in case the recipient never closes.
+    }
+
+
+    mapping (uint256 => PaymentChannel) public channels;
+    mapping (address => uint256)        public balances; //tokens which have been deposit but haven't been escrowed in the channels
+   
+    //already used messages for open_channel_by_recipient in order to prevent replay attack
+    mapping (bytes32 => bool) public used_messages; 
+
+    uint256 public next_channel_id; //id of the next channel (and size of channels)
+ 
+    ERC20 public token; // Address of token contract
+    
+    //TODO: optimize events. Do we need more (or less) events?
+    event event_channel_open       (uint256 channel_id,         address indexed sender, address indexed recipient, uint256 indexed replica_id);
```

Comment:
Looks like ```channel_id``` is enough, what are the benefits of having other information in event?

Reply:
We need "sender", "recipient" and "replica_id" here in order to simplify search of channels for the give client or for the given server. By the way we could add more events if needed, because they are relatively cheap...